File: go/execution/job/conditional.go
package job

import (
	"context"

	"github.com/gt-tech-ai/knowledge-engine/go/core/interfaces"
	"github.com/gt-tech-ai/knowledge-engine/go/core/types"
)

// Conditional wraps an AnyJob and runs it only when condition returns true.
// When the condition is false the job returns a single Skip result.
type Conditional struct {
	// inner is the wrapped job that runs when the condition holds.
	inner interfaces.AnyJob

	// condition is evaluated at execution time; the inner job runs only when it
	// returns true, otherwise the job reports a single Skip result.
	condition func(ctx context.Context, root string) bool
}

// Compile-time assertion that *Conditional satisfies interfaces.AnyJob.
var _ interfaces.AnyJob = (*Conditional)(nil)

// Meta implements interfaces.AnyJob.
func (c *Conditional) Meta() types.JobMeta { return c.inner.Meta() }

// Execute implements interfaces.AnyJob.
func (c *Conditional) Execute(
	ctx context.Context,
	runner interfaces.CommandRunner,
	root string,
) (types.StepResults, error) {
	if !c.condition(ctx, root) {
		return types.StepResults{
			{Name: c.inner.Meta().Name, Status: types.StatusSkip},
		}, nil
	}
	return c.inner.Execute(ctx, runner, root)
}

// When wraps job with a condition; the job runs only when condition returns true.
func When(
	job interfaces.AnyJob,
	condition func(ctx context.Context, root string) bool,
) *Conditional {
	return &Conditional{inner: job, condition: condition}
}

File: python/techai_webutils/src/techai_webutils/execution/job/conditional.py
"""Conditional job composition: run a wrapped AnyJob only when a predicate holds (port of conditional.go).

``when(job, predicate)`` yields a ``Conditional`` that runs ``job`` only when ``predicate()`` returns
True; otherwise it reports a single SKIP result named for the inner job, without running it.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from techai_webutils.core.interfaces.execution import BatchResult, StepResult, StepStatus

if TYPE_CHECKING:
    from collections.abc import Callable

    from techai_webutils.core.interfaces.execution import AnyJob, JobMeta


class Conditional:
    """Wraps an AnyJob and runs it only when its predicate holds (mirrors Go ``job.Conditional``)."""

    def __init__(self, inner: AnyJob, condition: Callable[[], bool]) -> None:
        """Wrap ``inner`` behind ``condition`` (evaluated at execute time)."""
        # inner is the wrapped job that runs when the condition holds.
        self._inner: AnyJob = inner
        # condition is evaluated at execute time; False short-circuits to a SKIP.
        self._condition: Callable[[], bool] = condition

    def meta(self) -> JobMeta:
        """Return the wrapped job's identity unchanged."""
        return self._inner.meta()

    async def execute(self) -> BatchResult:
        """Run the inner job when the condition holds, else return a single SKIP result."""
        if not self._condition():
            meta = self._inner.meta()
            return BatchResult(
                results=(StepResult(name=meta.name, group=meta.group, status=StepStatus.SKIP),),
            )
        return await self._inner.execute()


def when(job: AnyJob, condition: Callable[[], bool]) -> Conditional:
    """Wrap ``job`` with ``condition``; it runs only when the condition returns True (mirrors Go ``When``)."""
    return Conditional(job, condition)
